File: lab3-24.04.20/abandoned/controllers/GUIController.scala
package controllers

import java.awt.Desktop
import java.{lang, util}
import java.net.{URI, URL}
import java.util.ResourceBundle
import javafx.application.Platform
import javafx.beans.value.{ChangeListener, ObservableValue}
import javafx.collections.FXCollections
import javafx.scene.effect.DropShadow
import javafx.scene.paint.Color
import javafx.fxml.{FXML, FXMLLoader, Initializable}
import javafx.scene.Cursor
import javafx.scene.control.{ComboBox, TableView, TextArea, TextField}
import javafx.scene.input.MouseEvent
import javafx.scene.layout.{HBox, VBox}
import math.{MathFunction, InterpolationService}

/**
 * Управляет взаимодействием с пользователем посредством графического интерфейса.
 * @define RED_LIGHT эффект красного свечения по контуру объекта; может быть применён к объекту {@see Node}.
 * @define GREEN_LIGHT эффект красного свечения по контуру объекта; может быть применён к объекту {@see Node}.
 * @define BLUE_LIGHT эффект красного свечения по контуру объекта; может быть применён к объекту {@see Node}.
 * @author Артемий Кульбако.
 * @version 1.2
 */
class GUIController extends Initializable {

  //TODO: задать настоящие типы
  @FXML private var toolbar: HBox = _
  @FXML private var gControl: GraphController = _
  @FXML private var methodChooser: ComboBox[InterpolationService.SolveMethods] = _
  @FXML private var funcChooser: ComboBox[MathFunction] = _
  @FXML private var leftBoundInput: TextField = _
  @FXML private var rightBoundInput: TextField = _
  @FXML private var outputArea: TextArea = _
  @FXML private var pointsAmountFld: TextField = _
  @FXML private var pointsTable: TableView[Any] = _
  @FXML private var graphPlaceholder: VBox = _
  private val fxMethods = FXCollections.observableArrayList(InterpolationService.SolveMethods.NEWTON_POLYNOMIAL)
  private val fxEqs = FXCollections.observableArrayList(InterpolationService.equations)
  val RED_LIGHT = new DropShadow(25.0, 0.0, 0.0, Color.RED)
  val BLUE_LIGHT = new DropShadow(25.0, 0.0, 0.0, Color.DEEPSKYBLUE)
  val GREEN_LIGHT = new DropShadow(25.0, 0.0, 0.0, Color.LIGHTGREEN)
  private object Counter { var lineNumber = 0 }

  /**
   * Инициализирует экземпляр класс GUIController, навешивая на различные его элементы обработчики событий и эффекты
   * постобработки.
   * @param url адрес внешнего ресурса, который может быть загружен.
   * @param resourceBundle загруженный внешний ресурс, который может быть использован.
   */
  override def initialize(url: URL, resourceBundle: ResourceBundle): Unit = {
    toolbar.setOnMousePressed((_: MouseEvent) => { toolbar.setCursor(Cursor.MOVE) })
    toolbar.setOnMouseEntered((t: MouseEvent) => { if (!t.isPrimaryButtonDown) toolbar.setCursor(Cursor.HAND) })
    toolbar.setOnMouseExited((t: MouseEvent) => { if (!t.isPrimaryButtonDown) toolbar.setCursor(Cursor.DEFAULT) })
    Seq(methodChooser, funcChooser, leftBoundInput, rightBoundInput, pointsAmountFld, pointsTable).foreach(it => {
      it.focusedProperty().addListener(new ChangeListener[lang.Boolean] {
        override def changed(obsVal: ObservableValue[_ <: lang.Boolean], oldVal: lang.Boolean, newVal: lang.Boolean): Unit = {
          it.setEffect(if (newVal) BLUE_LIGHT else null)
        }})
    })
    Seq(leftBoundInput, rightBoundInput).foreach(it => {
      it.textProperty().addListener(new ChangeListener[String] {
        override def changed(obsVal: ObservableValue[_ <: String], oldVal: String, newVal: String): Unit = {
          if (!newVal.matches("-?\\d{0,2}([.]\\d{0,6})?"))
            printMessage(RED_LIGHT, "Поля должны быть представлены числом", "Максимальная точность - 6 знаков после запятой")
        }})
    })
    pointsAmountFld.textProperty().addListener(new ChangeListener[String] {
      override def changed(obsVal: ObservableValue[_ <: String], oldVal: String, newVal: String): Unit = {
        if (!newVal.matches("^$|([1-9]|1[013])$"))
          printMessage(RED_LIGHT, "Поле должно быть целым числом, от 1 до 13")
      }
    })
    methodChooser.setItems(fxMethods)
    funcChooser.setItems(fxEqs)
    val loader = new FXMLLoader(getClass.getResource("/resources/graph.fxml"))
    graphPlaceholder.getChildren.add(loader.load())
    gControl = loader.getController.asInstanceOf[GraphController]
  }

  /** Минимизирует окно программы.*/
  @FXML def minimizeWindow(): Unit = { AeroMain.stage.setIconified(true) }

  /** Завершает работу программы.*/
  @FXML def closeProgram(): Unit = Platform.exit()

  /** Открывает профиль разработчика на github через установленный по-умолчанию веб-браузер в системе.*/
  @FXML def showCredits(): Unit = Desktop.getDesktop.browse(new URI("https://github.com/testpassword"))

  /** Печатает сообщение в главное окно вывода. */
  @FXML def printMessage(lightning: DropShadow, strings: String*): Unit = {
    strings.foreach(it => {
      outputArea.appendText(s"${Counter.lineNumber}. $it\n")
      Counter.lineNumber += 1
    })
    outputArea.setEffect(lightning)
  }

  @FXML private def showResult(): Unit = {
    try {
      val method = methodChooser.getValue
      val f = funcChooser.getValue
      val borders = (leftBoundInput.getText.toDouble, rightBoundInput.getText.toDouble)
      val n = pointsAmountFld.getText.toInt
      val res = InterpolationService.solve()
      gControl.clear()
      gControl.drawLine(f)
      /*
      Нарисовать точки
      Заполнить таблицу x и y, количество строк = n

      Пользователь может подправить точки в таблице, и сгенерировать график заново
      После генер. доступа вычисление в любой
      После подправления заблокировать выч. в любой точке
       */
    } catch {
      case _: NumberFormatException => printMessage(RED_LIGHT, "Одно из полей не заполнено или заполнено неверно",
        "Читай подсказки при вводе данных")
      case e: Throwable => printMessage(RED_LIGHT, e.getLocalizedMessage)
    }
  }
}

File: lab4-29.05.20/src/controllers/GUIController.kt
package controllers

import AeroMain
import javafx.application.Platform
import javafx.collections.FXCollections
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.fxml.Initializable
import javafx.geometry.Point2D
import javafx.scene.Cursor
import javafx.scene.control.*
import javafx.scene.effect.DropShadow
import javafx.scene.input.MouseEvent
import javafx.scene.layout.BorderPane
import javafx.scene.layout.HBox
import javafx.scene.paint.Color
import math.MathFunction
import math.differential.OrdinaryDifferentialSolver
import math.interpolation.InterpolationSolver
import org.mariuszgromada.math.mxparser.Argument
import org.mariuszgromada.math.mxparser.Expression
import java.awt.Desktop
import java.net.*
import java.util.*
import kotlin.Exception

/**
 * Управляет взаимодействием с пользователем посредством графического интерфейса.
 * @property RED_LIGHT эффект красного свечения по контуру объекта; может быть применён к объекту {@see Node}.
 * @property GREEN_LIGHT эффект красного свечения по контуру объекта; может быть применён к объекту {@see Node}.
 * @property BLUE_LIGHT эффект красного свечения по контуру объекта; может быть применён к объекту {@see Node}.
 * @author Артемий Кульбако.
 * @version 1.5
 */
class GUIController: Initializable {

    companion object {
        val RED_LIGHT = DropShadow(25.0, 0.0, 0.0, Color.RED)
        val BLUE_LIGHT = DropShadow(25.0, 0.0, 0.0, Color.DEEPSKYBLUE)
        val GREEN_LIGHT = DropShadow(25.0, 0.0, 0.0, Color.LIGHTGREEN)
        private var linesCounter = 1
    }
    @FXML private lateinit var toolbar: HBox
    @FXML private lateinit var gControl: GraphController
    @FXML private lateinit var methodChooser: ComboBox<OrdinaryDifferentialSolver>
    @FXML private lateinit var funcInput: TextField
    @FXML private lateinit var leftBoundInput: TextField
    @FXML private lateinit var rightBoundInput: TextField
    @FXML private lateinit var leftBoundValInput: TextField
    @FXML private lateinit var accInput: TextField
    @FXML private lateinit var outputArea: TextArea
    @FXML private lateinit var mainPane: BorderPane
    @FXML private lateinit var interpolationControl: HBox
    @FXML private lateinit var xInput: TextField
    @FXML private lateinit var yOutput: TextField
    private lateinit var method: OrdinaryDifferentialSolver
    private lateinit var approxFunc: MathFunction<Double>
    private var dots: List<Point2D>? = null

    /**
     * Инициализирует экземпляр класс [GUIController], навешивая на различные его элементы обработчики событий и эффекты
     * постобработки.
     * @param url адрес внешнего ресурса, который может быть загружен.
     * @param bundle внешний ресурс, который может быть использован.
     */
    override fun initialize(url: URL?, bundle: ResourceBundle?) {
        toolbar.let {
            it.onMousePressed = EventHandler { mouseEvent: MouseEvent -> it.cursor = Cursor.MOVE }
            it.onMouseEntered = EventHandler { mouseEvent: MouseEvent -> if (!mouseEvent.isPrimaryButtonDown) { it.cursor = Cursor.HAND } }
            it.onMouseExited = EventHandler { mouseEvent: MouseEvent -> if (!mouseEvent.isPrimaryButtonDown) { toolbar.cursor = Cursor.DEFAULT } }
        }
        sequenceOf(methodChooser, funcInput, leftBoundInput, rightBoundInput, leftBoundValInput, accInput, xInput,
            yOutput).forEach {
            it.focusedProperty().addListener { _, _, newVal ->
                it.effect = if (newVal!!) BLUE_LIGHT else null }
        }
        val resetDots = {
            dots?.let {
                dots = null
                printMessage(BLUE_LIGHT, "Точки сброшены, так как параметры функции были изменены")
            }
            interpolationControl.isDisable = true
        }
        sequenceOf(methodChooser, funcInput, leftBoundInput, rightBoundInput, leftBoundValInput, accInput).forEach {
            when (it) {
                is ComboBox<out Any> -> it.valueProperty().addListener {_, _, _ -> resetDots() }
                is TextField -> it.textProperty().addListener {_, _, _ -> resetDots() }
            }
        }
        sequenceOf(leftBoundInput, rightBoundInput, leftBoundValInput, xInput).forEach {
            it.textProperty().addListener{ _, oldVal, newVal ->
                if (!newVal.matches(Regex("-?\\d{0,2}([.]\\d{0,6})?"))) {
                    printMessage(RED_LIGHT, "Поле должны быть представлены числом", "Максимальная точность - 6 знаков после запятой")
                    println(oldVal)
                    it.text = oldVal
                }
            }
        }
        xInput.textProperty().addListener { _, oldVal, newVal ->
            val x = newVal.toDouble()
            val min = dots!!.minBy { it.x }!!.x
            val max = dots!!.maxBy { it.x }!!.x
            if (x in min..max) {
                val y = approxFunc.func(x)
                yOutput.text = y.toString()
                val p = Point2D(x, y)
                redrawGraph(Pair(min, max), p)
                printMessage(GREEN_LIGHT, "Значение функции в f'(x) = $y")
            } else {
                printMessage(RED_LIGHT, "x должно быть в интервале [${leftBoundInput.text} ; ${rightBoundInput.text})")
                xInput.text = oldVal?: ""
            }
        }
        methodChooser.valueProperty().addListener {
                _, _, _ -> method = methodChooser.value }
        methodChooser.items = FXCollections.observableList(OrdinaryDifferentialSolver.values().toList())
        FXMLLoader(javaClass.getResource("/resources/graph.fxml")).let {
            mainPane.right = it.load()
            gControl = it.getController() as GraphController
        }
    }

    /** Минимизирует окно программы.*/
    @FXML fun minimizeWindow() { AeroMain.stage!!.isIconified = true }

    /** Завершает работу программы.*/
    @FXML fun closeProgram() = Platform.exit()

    /** Открывает профиль разработчика на github через стандартный веб-браузер в системе.*/
    @FXML fun showCredits() = Desktop.getDesktop().browse(URI("https://github.com/testpassword"))

    /** Печатает сообщение в главное окно вывода. */
    fun printMessage(lightning: DropShadow, vararg strings: String) {
        strings.forEach {
            outputArea.appendText("$linesCounter. $it\n")
            linesCounter++
        }
        outputArea.effect = lightning
    }

    private fun redrawGraph(borders: Pair<Double, Double>, extraPoint: Point2D? = null) =
        gControl.let {
            it.clear()
            it.drawLine(this.approxFunc, borders)
            dots!!.forEach {p -> it.drawPoint(p) }
            if (extraPoint != null) gControl.drawPoint(extraPoint, Color.DARKKHAKI)
        }

    @FXML private fun showResult() {
        try {
            val expression = Expression(funcInput.text).apply {
                this.addArguments(Argument("x"), Argument("y"))
                if (!this.checkSyntax()) throw IllegalArgumentException("Уравнение введено неверно")
            }
            val x0 = leftBoundInput.text.toDouble()
            val xn = rightBoundInput.text.toDouble()
            val y0 = leftBoundValInput.text.toDouble()
            if (x0 >= xn) throw Exception("x0 =$x0 должно быть меньше xn=$xn")
            val accuracy = accInput.text.toDouble().let {
                if (it <= 0) {
                    printMessage(RED_LIGHT, "Точность должна быть строго больше 0", "Мы поправили её за вас 🥰")
                    it.coerceAtLeast(0.01)
                } else it
            }
            if (((xn - x0) / accuracy) > 50) throw IllegalArgumentException("Сильно много точек, JavaFX не поймёт 😭")
            this.dots = OrdinaryDifferentialSolver.EULER.solve(expression, Point2D(x0, y0), accuracy, (x0..xn))
            this.approxFunc = InterpolationSolver.newtonPolynomial(this.dots!!)
            redrawGraph(Pair(x0, xn))
            interpolationControl.isDisable = false
            printMessage(GREEN_LIGHT, "Решение ОДУ найдено")
        } catch (e: Exception) {
            when (e) {
                is NumberFormatException -> printMessage(RED_LIGHT, "Одно из полей незаполненно или заполнено неверно")
                else -> printMessage(RED_LIGHT, e.localizedMessage)
            }
        }
    }
}